backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/CloneVmCommand.java

```diff
@@ -134,13 +136,19 @@ protected void init() {
 
     @Override
     protected void executeVmCommand() {
-        getParameters().setStage(CloneVmParameters.CloneVmStage.CREATE_VM_SNAPSHOT);
-        Guid snapshotId = createVmSnapshot();
-        setSnapshotId(snapshotId);
-        setSucceeded(snapshotId != null);
+        if (isNoManagedBlockDiskAttachedToVm()) {
+            getParameters().setStage(CloneVmParameters.CloneVmStage.CREATE_VM_SNAPSHOT);
+            Guid snapshotId = getVmSnapshotToClone();
+            initDisks(snapshotId);
+            setSucceeded(snapshotId != null);
+        } else {
+            getParameters().setStage(CloneVmParameters.CloneVmStage.COPY_DISKS);
+            setSucceeded(true);
+        }
```

Comment:
So you add special flow for MBS disk. This is not a good design. Soon we will have here a branch for every kind of storage, and it will be quickly unmaintainable.

If the right solution is not to create a snapshot for MBS (not sure about this)
we need to covert this to capability of the disk or storage domain that can be
queried by other code without depending on specific storage type.

For example, if the capability is being able to copy a snapshot, we can have
something like this:

    if (storageDomain.canCopySnapshot()) {
        flow using snapshot...
    } else {
        flow not using snapshots...
    }

Introducing lot of conditions like this in the code is not that great, but it is 
much better than depending on specific storage type.

Reply:
`I'm not sure what is the "future" of snapshots for MBS disks.`
The future is bright :)

I mean, the priority in my opinion, is at the very least to restore what was possible originally, which is to clone VM using the clone_volume cinderlib verb, without snapshots on the same storage domain.

Since clone VM has significantly changed and now supports cloning to different storage domains and works live, we'd need to see how MBS can catch up with these changes and how important it is.

While there is merit to having feature gates in the form `storageDomain.canCopySnapshot`, we don't have this abstraction. What we have is https://github.com/oVirt/ovirt-engine/blob/master/backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/validator/storage/ManagedBlockStorageDomainValidator.java which blocks entire operations based on the capabilities of the storage domain type, but we don't have something in the granularity of changing how an operation is performed based on the storage domain type.

Reply:
Since this code is "closed" to the original code, I think it will be easier to track currently.
@nirs I get your fair point, about the code's durability and maintaining it in the future, considering other storage domain's types, but I don't think it's that easy to "make it happen", and might be out of the scope of this patch,
as I personally think we should focus on fixing the broken behavior we currently have.
Regarding your suggestion, I think it's a bit complex since you need to overview the VMs disks, and by that set the coco phases (as shown in the patch), so I think that going over the VM disks and see if there's at least one MBS disk is more "straight-forward", so after advising with @bennyz, I think we can stay with this one.

@smelamud, I tried to keep the code as much the same as you wrote it, but there are some adjustments for MBS disks, and taking them into account before snapshotting the VM.
Can you please share your feedback?

Reply:
a friendly ping? :)

Reply:
The code of `CloneVmCommand` is already overly complicated and introducing a more sophisticated logic to it may be harder than it seems. @sleviim, I think it would be better for now to go forward with the approach you've proposed. Just rename `isNoManagedBlockDiskAttachedToVm` to something more generic like `isAllVmDisksSupportSnapshots` - this way it will be clear in the future what capabilities should be checked here.

Reply:
Ack, thanks.